haystack/core/pipeline/base.py

```diff
@@ -1042,24 +1042,134 @@ def _add_missing_input_defaults(component_inputs: Dict[str, Any], component_inpu
 
         return component_inputs
 
-    @staticmethod
+    def _tiebreak_waiting_components(
+        self,
+        component_name: str,
+        priority: ComponentPriority,
+        priority_queue: FIFOPriorityQueue,
+        topological_sort: Union[Dict[str, int], None],
+    ):
+        """
+        Decides which component to run when multiple components are waiting for inputs with the same priority.
+        """
+        components_with_same_priority = [component_name]
+
+        while len(priority_queue) > 0:
+            next_priority, next_component_name = priority_queue.peek()
+            if next_priority == priority:
+                priority_queue.pop()  # actually remove the component
+                components_with_same_priority.append(next_component_name)
+            else:
+                break
+
+        if len(components_with_same_priority) > 1:
+            if topological_sort is None:
+                if networkx.is_directed_acyclic_graph(self.graph):
+                    topological_sort = networkx.lexicographical_topological_sort(self.graph)
+                    topological_sort = {node: idx for idx, node in enumerate(topological_sort)}
+                else:
+                    condensed = networkx.condensation(self.graph)
+                    condensed_sorted = {node: idx for idx, node in enumerate(networkx.topological_sort(condensed))}
```

Comment:
More for my own education, why not use the condensation call for both?

Reply:
Interesting point. It's not needed for a DAG but it should still work and we could skip the if/else.

What I'm not sure about are the performance implications. Since `condensation` is creating a whole new graph, I could imagine that it is more expensive to run than `is_directed_acyclic_graph`.

Reply:
Okay no worries let's leave as is. I'd rather use the specific methods when we can, just more for me to understand networkx a bit more.